lib/pure/options.nim

```diff
@@ -176,22 +297,240 @@ proc filter*[T](self: Option[T], callback: proc (input: T): bool): Option[T] =
   else:
     self
 
-proc `==`*(a, b: Option): bool =
-  ## Returns ``true`` if both ``Option``s are ``none``,
-  ## or if they have equal values
-  (a.isSome and b.isSome and a.val == b.val) or (not a.isSome and not b.isSome)
-
 proc `$`*[T](self: Option[T]): string =
   ## Get the string representation of this option. If the option has a value,
-  ## the result will be `Some(x)` where `x` is the string representation of the contained value.
-  ## If the option does not have a value, the result will be `None[T]` where `T` is the name of
-  ## the type contained in the option.
+  ## the result will be `Some(x)` where `x` is the string representation of the
+  ## contained value. If the option does not have a value, the result will be
+  ## `None[T]` where `T` is the name of the type contained in the option.
   if self.isSome:
     "Some(" & $self.val & ")"
   else:
     "None[" & name(T) & "]"
 
+macro `.?`*(option: untyped, statements: untyped): untyped =
+  ## Optional continuation operator. Works like regular dot-chaining, but if
+  ## the left had side is a ``none`` then the right hand side is not evaluated.
+  ## In the case that ``statements`` return something the return type of this
+  ## will be ``Option[T]`` where ``T`` is the returned type of ``statements``.
+  ## If nothing is returned from ``statements`` this returns nothing.
+  ##
+  ## .. code-block:: nim
+  ##   echo some("Hello").?find('l') ## Prints out Some(2)
+  ##   some("Hello").?find('l').echo # Prints out 2
+  ##   none(string).?find('l').echo # Doesn't print out anything
+  ##   echo none(string).?find('l') # Prints out None[int] (return type of find)
+  ##   # These also work in things like ifs
+  ##   if some("Hello").?find('l') == 2:
+  ##     echo "This prints"
+  ##   if none(string).?find('l') == 2:
+  ##     echo "This doesn't"
+  let opt = genSym(nskLet)
+  var
+    injected = statements
+    firstBarren = statements
+  if firstBarren.len != 0:
+    while true:
+      if firstBarren[0].len == 0:
+        firstBarren[0] = nnkDotExpr.newTree(
+          nnkDotExpr.newTree(opt, newIdentNode("val")), firstBarren[0])
+        break
+      firstBarren = firstBarren[0]
+  else:
+    injected = nnkDotExpr.newTree(
+      nnkDotExpr.newTree(opt, newIdentNode("val")), firstBarren)
+
+  result = quote do:
+    (proc (): auto =
+      let `opt` = `option`
+      if `opt`.isSome:
+        when compiles(`injected`) and not compiles(some(`injected`)):
+          `injected`
+        else:
+          return some(`injected`)
+    )()
+
+  #echo result.treerepr
+
+macro match*(option: untyped, body: untyped): untyped =
```

Comment:
I'm not really comfortable with this name - I'd much prefer something like optionCase instead.
- `match` is a very generic name, that doesn't indicate what it does very well.
- Using two untyped parameters means that it can't really be overloaded, for other libraries that use 'match'.

Reply:
Yeah as I've said I'm up for changing the name. Just changed it to match since people seemed to prefer that. `optionCase` might be a bit long though, but it's very descriptive.